### PR TITLE
[NT-2105] Mutation Create Payment Source

### PR DESCRIPTION
**Pull request by msadoon**
# 📲 What

Replaced the existing `addNewCreditCard` endpoint to use Apollo.

# 🤔 Why

Convert GQL Schema use to Apollo.

# 🛠 How

- Added `CreatePaymentSource.graphql` mutation
- Added adapters `CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift`, `GraphAPI.CreatePaymentSourceInput+CreatePaymentSourceInput.swift`
- Renamed `GraphUserCreditCard` to `UserCreditCards` because it contains a property called `storedCards` and `GraphUserCreditCard.storedCards` seemed weird. Also we are moving away from the `Graph` prefix on objects to demote use of the GQL schema. Over 90% of the changes in this PR are this renaming change.

# ✅ Acceptance criteria

- [x] Can create a "remembered" and "non-remembered" payment method from the pledge checkout screen.
- [x] Can create a "remembered" payment method from the settings payment method screen.
- [ ] Ensure if creating a payment method fails. An appropriate error message is shown.

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: KsApi/models/CreatePaymentSourceEnvelope.swift [hunk 1]
```diff
@@ -4,19 +4,7 @@ public struct CreatePaymentSourceEnvelope: Decodable {
...
-
-extension CreatePaymentSourceEnvelope {
-  private enum CodingKeys: String, CodingKey {
-    case createPaymentSource
-  }
-
-  public init(from decoder: Decoder) throws {
-    let values = try decoder.container(keyedBy: CodingKeys.self)
-    self.createPaymentSource = try values.decode(CreatePaymentSource.self, forKey: .createPaymentSource)
+    public var paymentSource: UserCreditCards.CreditCard
```
Comment (singhhari) [inline KsApi/models/CreatePaymentSourceEnvelope.swift:8]: Why did you remove the `errorMessage` property?

File: KsApi/models/DeletePaymentMethodEnvelope.swift [hunk 2]
```diff
@@ -17,6 +17,6 @@ extension DeletePaymentMethodEnvelope: Decodable {
     self.storedCards = try values.nestedContainer(keyedBy: CodingKeys.self, forKey: .paymentSourceDelete)
       .nestedContainer(keyedBy: CodingKeys.self, forKey: .user)
       .nestedContainer(keyedBy: CodingKeys.self, forKey: .storedCards)
-      .decode([GraphUserCreditCard.CreditCard].self, forKey: .nodes)
+      .decode([UserCreditCards.CreditCard].self, forKey: .nodes)
```
Comment (singhhari) [inline KsApi/models/DeletePaymentMethodEnvelope.swift:20]: Not sure why this is being flagged as untested, I've run the tests locally and see the coverage here.

File: KsApi/models/graphql/UserCreditCardsTests.swift [hunk 3]
```diff
@@ -12,11 +12,11 @@ final class GraphUserCreditCardTests: XCTestCase {
     let data = jsonString.data(using: .utf8)
 
     do {
-      let cards = try JSONDecoder().decode(GraphUserCreditCard.self, from: data!)
+      let cards = try JSONDecoder().decode(UserCreditCards.self, from: data!)
 
       XCTAssertEqual(cards.storedCards.count, 0)
     } catch {
-      XCTFail("Failed to decode GraphUserCreditCard")
+      XCTFail("Failed to decode UserCreditCards")
```
Comment (singhhari) [inline KsApi/models/graphql/UserCreditCardsTests.swift:19]: Also shouldn't be flagging a test file here. Can you look into this and see why codecov is flagging errantly?

File: KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import Foundation
+import ReactiveSwift
+
+extension CreatePaymentSourceEnvelope {
+  static func from(_ data: GraphAPI.CreatePaymentSourceMutation.Data) -> CreatePaymentSourceEnvelope? {
+    guard let createdPaymentSource = data.createPaymentSource,
+      createdPaymentSource.isSuccessful,
```
Comment (singhhari) [inline KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift:7]: Curious if we need to perform this check here?

File: KsApi/models/graphql/adapters/GraphAPI.CreatePaymentSourceInput+CreatePaymentSourceInput.swift [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+extension GraphAPI.CreatePaymentSourceInput {
+  static func from(_ input: CreatePaymentSourceInput) -> GraphAPI.CreatePaymentSourceInput {
+    return GraphAPI.CreatePaymentSourceInput(
+      paymentType: .creditCard,
```
Comment (singhhari) [inline KsApi/models/graphql/adapters/GraphAPI.CreatePaymentSourceInput+CreatePaymentSourceInput.swift:6]: Is creditCard really the only payment source we accept? I wonder if debit cards are interpreted as credit card in this case?

**Review by singhhari — COMMENTED** [review]
nice job. lgtm.

prob good to merge here but left a few questions for my own information.

File: KsApi/models/CreatePaymentSourceEnvelope.swift (hunk 1 above)
Comment (msadoon) [reply KsApi/models/CreatePaymentSourceEnvelope.swift:8]: The `errorMessage` property on the GQL mutation is deprecated, so no need to reference it here or in the mutation anymore.

File: KsApi/models/DeletePaymentMethodEnvelope.swift (hunk 2 above)
Comment (msadoon) [reply KsApi/models/DeletePaymentMethodEnvelope.swift:20]: code coverage on xcode (Editor --> Code Coverage) shows this isn't currently being covered, but it's not a file related to this ticket, it is one of the files covered by the next ticket this sprint. I just had to update the name reference here.

Comment (singhhari) [reply KsApi/models/DeletePaymentMethodEnvelope.swift:20]: I think codecov isn't managed by xcode but rather it's website or yaml file btw

File: KsApi/models/graphql/UserCreditCardsTests.swift (hunk 3 above)
Comment (msadoon) [reply KsApi/models/graphql/UserCreditCardsTests.swift:19]: Ok, previously the `.codecov.yml` didn't ignore any files, my hunch is that adding `*Tests.swift` to the top level header `.ignore` will remove reporting coverage for the test files themselves. I will push that commit up shortly.

Comment (msadoon) [reply KsApi/models/graphql/UserCreditCardsTests.swift:19]: Let's see if that clears these warnings.

File: KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift (hunk 4 above)
Comment (msadoon) [reply KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift:7]: yea, there's no description of this flag on GraphiQL, I just assumed it meant the mutation was successful. Seemed to be returning `true` for all mutations I tried. Presumably if a mutation fails on backend, this flag is `true`. It was being parsed in the GQL schema, so I thought might as well use it here.

Comment (msadoon) [reply KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift:7]: I can remove it though.

Comment (msadoon) [reply KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift:7]: I can remove it, mostly just used it for testing.

File: KsApi/models/graphql/adapters/GraphAPI.CreatePaymentSourceInput+CreatePaymentSourceInput.swift (hunk 5 above)
Comment (msadoon) [reply KsApi/models/graphql/adapters/GraphAPI.CreatePaymentSourceInput+CreatePaymentSourceInput.swift:6]: Yea I know it was strange, but the `GraphAPI` only supports `.creditCard` according to `line 367` on `GraphAPI`.

File: KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift (hunk 4 above)
Comment (msadoon) [reply KsApi/models/graphql/adapters/CreatePaymentSourceEnvelope+CreatePaymentSourceMutation.Data.swift:7]: Oops, I meant to say it's false if the query fails, I remembered trying invalid data on a mutation here (ie. tokenId was empty string) returned this flag as `false`. So it is useful for any reason the backend didn't make the requested change. Can be like a early exit before trying to parse all of the data.

**Review by singhhari — APPROVED** [review]
lgtm

File: KsApi/models/graphql/UserCreditCardsTests.swift (hunk 3 above)
Comment (msadoon) [reply KsApi/models/graphql/UserCreditCardsTests.swift:19]: No warnings seem to appear on test files anymore and the coverage didn't drop, so that's a good sign.

Comment (msadoon) [reply KsApi/models/graphql/UserCreditCardsTests.swift:19]: Oh wait this is still here.

**Comment by msadoon** [comment]
So I think the default patch code coverage comes from [here](https://app.codecov.io/gh/kickstarter/ios-oss/settings/yaml), not our local `codecov.yaml`. Not sure why that is, but I'll revert the change I made to `.codecov.yml` in my last commit before merging - but I don't seem to have access rights to change the above.
